xdsl/dialects/mpi.py

```diff
@@ -352,4 +356,461 @@ class Finalize(MPIBaseOp):
+
+    MPI_INT: int = 0x4c000405
+    MPI_UNSIGNED: int = 0x4c000406
+    MPI_LONG: int = 0x4c000807
+    MPI_UNSIGNED_LONG: int = 0x4c000808
+    MPI_FLOAT: int = 0x4c00040a
+    MPI_DOUBLE: int = 0x4c00080b
+    MPI_UNSIGNED_CHAR: int = -1
+    MPI_UNSIGNED_SHORT: int = -1
+    MPI_UNSIGNED_LONG_LONG: int = -1
+    MPI_CHAR: int = -1
+    MPI_SHORT: int = -1
+    MPI_LONG_LONG_INT: int = -1
+
+    MPI_STATUS_IGNORE: int = 1
+
+    request_size: int = 4
+    status_size: int = 4 * 5
+    mpi_comm_size: int = 4
+
+
+class MpiLowerings(RewritePattern):
+    """
+    This rewrite pattern contains rules to lower the MPI dialect to llvm+func+builin
+
+    In order to lower that far, we require some information about the targeted MPI library
+    (magic values, struct sizes, field offsets, etc.). This information is provided using
+    the MpiLibraryInfo class.
+    """
+
+    _emitted_function_calls: dict[str, tuple[list[Attribute], list[Attribute]]]
+    """
+    This object keeps track of all the functions we have emitted and their type signature.
+
+    This is done so we can later on add "external" function declarations so LLVM is happy :)
+    """
+
+    MPI_SYMBOL_NAMES = {
+        'mpi.init': 'MPI_Init',
+        'mpi.finalize': 'MPI_Finalize',
+        'mpi.irecv': 'MPI_Irecv',
+        'mpi.isend': 'MPI_Isend',
+        'mpi.wait': 'MPI_Wait',
+        'mpi.comm.rank': 'MPI_Comm_rank',
+        'mpi.recv': 'MPI_Recv',
+        'mpi.send': 'MPI_Send'
+    }
+    """
+    Translation table for mpi operation names to their MPI library function names
+    """
+
+    info: MpiLibraryInfo
+    """
+    This object carries information about the targeted MPI libraray
+    """
+
+    def __init__(self, info: MpiLibraryInfo):
+        self.info = info
+        self._emitted_function_calls = dict()
+
+    def match_and_rewrite(self, op: Operation, rewriter: PatternRewriter):
+        """
+        This method acts as a dispatcher to lower individual MPI operations.
+
+        It tries to dispatch calls for each mpi.<name> op to lower_mpi_<name> methods.
+
+        The methods each return the argument inputs to rewriter.replace_matched_op calls
+        """
+        if not isinstance(op, MPIBaseOp):
+            return
+
+        field = "lower_{}".format(op.name.replace('.', '_'))
+
+        if hasattr(self, field):
+            new_ops, *other = getattr(self, field)(op)
+            rewriter.replace_matched_op(new_ops, *other)
+
+            for op in new_ops:
+                if not isinstance(op, func.Call):
+                    continue
+                if not op.callee.string_value().startswith('MPI_'):
+                    continue
+                self._emitted_function_calls[op.callee.string_value()] = (
+                    [x.typ for x in op.arguments],
+                    [x.typ for x in op.results],
+                )
+        else:
+            print("Missing lowering for {}".format(op.name))
+
+    # Individual lowerings:
```

Comment:
Each of these should be their own `match_and_rewrite`

Reply:
That is repeating so much code though, isn't it a very common pattern in rewriting that you have one pass over the entire program and you replace all the operations you match?

I feel that having individual `RewritePattern` classes for each operation is introducing a lot of code, which in turn introduces the ability for lots of bugs if you miss stuff, right?

One of my main principles in coding is to reduce the number of places you need to edit stuff for each addition of a single feature.

Reply:
Usually, a pass consists of many patterns to be applied. Also, a lot of the walker logic you have here is not actually needed, if you use our PatternRewriter. You can also parametrize that walker to walk inside-out (regions first) or in-reverse, so there is advantages in using that, IMO.

I think your approach works, but I am a bit hesitant to merge it as it does not really reflect the way the rewrite should be used, so I don't want to merge it as it would be a bad example, IMO.

Reply:
If you dont have shared state between rewrites, you could implement this with anonymous rewrites, which means even less code than you have at the moment.

Reply:
I updated it!